dockstore-webservice/src/main/java/io/dockstore/webservice/resources/MetadataResource.java

```diff
@@ -120,30 +130,48 @@ public MetadataResource(SessionFactory sessionFactory, DockstoreWebserviceConfig
     @Operation(summary = "List all available workflow, tool, organization, and collection paths.", description = "List all available workflow, tool, organization, and collection paths. Available means published for tools/workflows, and approved for organizations and their respective collections. NO authentication")
     @ApiOperation(value = "List all available workflow, tool, organization, and collection paths.", notes = "List all available workflow, tool, organization, and collection paths. Available means published for tools/workflows, and approved for organizations and their respective collections.")
     public String sitemap() {
-        //TODO needs to be more efficient via JPA query
-        List<Tool> tools = toolDAO.findAllPublished();
-        // #2683 avoid web crawling services for 1.7.0
-        List<Workflow> workflows = workflowDAO.findAllPublished(null, Integer.MAX_VALUE, null, null, null, (Class)BioWorkflow.class);
+        List<String> urls = new ArrayList<>();
+        addToolPaths(urls);
+        addBioWorkflowPaths(urls);
+        // Do not append services yet
+        // addServicePaths(urls);
+        addOrganizationAndCollectionPaths(urls);
+        Collections.sort(urls);
+        return String.join(System.lineSeparator(), urls);
+    }
+
+    /**
+     * Adds organization and collection URLs
+     * //TODO needs to be more efficient via JPA query
+     * @param urls  Current list of all URLs for sitemap
+     */
+    private void addOrganizationAndCollectionPaths(List<String> urls) {
         List<Organization> organizations = organizationDAO.findAllApproved();
-        StringBuilder builder = new StringBuilder();
-        for (Tool tool : tools) {
-            builder.append(createToolURL(tool));
-            builder.append(System.lineSeparator());
-        }
-        for (Workflow workflow : workflows) {
-            builder.append(createWorkflowURL(workflow));
-            builder.append(System.lineSeparator());
-        }
-        for (Organization organization: organizations) {
-            builder.append(createOrganizationURL(organization));
-            builder.append(System.lineSeparator());
+        organizations.forEach(organization -> {
+            urls.add(createOrganizationURL(organization));
             List<Collection> collections = collectionDAO.findAllByOrg(organization.getId());
-            for (Collection collection: collections) {
-                builder.append(createCollectionURL(collection, organization));
-                builder.append(System.lineSeparator());
-            }
-        }
-        return builder.toString();
+            collections.stream().map(collection -> createCollectionURL(collection, organization)).forEach(urls::add);
+        });
+    }
+
+    private void addToolPaths(List<String> urls) {
```

Comment:
The nascent functional side of me would instead have these functions return `List<String>` and have the caller use the return value, e.g., up above have `urls.addAll(addToolPaths())` instead of this side effect of modifying a parameter.

Feel free to ignore this comment.

Reply:
I prefer no side-effects as well